sdks/go/pkg/beam/core/runtime/exec/datasource.go

```diff
@@ -200,5 +230,41 @@ func (n *DataSource) Progress() ProgressReportSnapshot {
 	if n == nil {
 		return ProgressReportSnapshot{}
 	}
-	return ProgressReportSnapshot{n.SID.Target.ID, n.SID.Target.Name, atomic.LoadInt64(&n.count)}
+	n.mu.Lock()
+	c := n.count
+	n.mu.Unlock()
+	return ProgressReportSnapshot{n.SID.Target.ID, n.SID.Target.Name, c}
+}
+
+// Split takes a sorted set of potential split points, selects and actuates
+// split on an appropriate split point, and returns the selected split point
+// if successful. Returns an error when unable to split.
+func (n *DataSource) Split(splits []int64, frac float32) (int64, error) {
+	if splits == nil {
+		return 0, fmt.Errorf("failed to split: requested splits were empty")
+	}
+	if n == nil {
+		return 0, fmt.Errorf("failed to split at requested splits: {%v}, DataSource not initialized", splits)
+	}
+	n.mu.Lock()
+	c := n.count
```

Comment:
Since we're in the "locked" zone here, we should probably avoid copying the value, for readability.

Reply:
I'm also using 'c' outside the lock to print the value that we saw when we held the lock...

Reply:
I missed that! Thanks!